zebra-rpc/src/methods/get_block_template_rpcs.rs

```diff
@@ -333,40 +333,113 @@ where
-            check_synced_to_tip(network, latest_chain_tip, sync_status)?;
-
-            // Fetch the state data and local time for the block template:
-            // - if the tip block hash changes, we must return from long polling,
-            // - if the local clock changes on testnet, we might return from long polling
             //
-            // We also return after 90 minutes on mainnet, even if we have the same response.
-            let chain_tip_and_local_time = fetch_state_tip_and_local_time(state).await?;
+            // Set up the loop.
+            let mut max_time_reached = false;
+
+            // The loop returns the server long poll ID,
+            // which should be different to the client long poll ID.
+            let (server_long_poll_id, chain_tip_and_local_time, mempool_txs) = loop {
+                // Check if we are synced to the tip.
+                // The result of this check can change during long polling.
+                //
+                // TODO:
+                // - add `async changed()` methods to ChainTip and ChainSyncStatus
+                //   (using `changed() -> Changed` and `impl Future<()> for Changed`)
+                check_synced_to_tip(network, latest_chain_tip.clone(), sync_status.clone())?;
+
+                // Fetch the state data and local time for the block template:
+                // - if the tip block hash changes, we must return from long polling,
+                // - if the local clock changes on testnet, we might return from long polling
+                //
+                // We always return after 90 minutes on mainnet, even if we have the same response,
+                // because the max time has been reached.
+                //
+                // TODO: timeout and exit the loop when max time is reached
+                let chain_tip_and_local_time =
+                    fetch_state_tip_and_local_time(state.clone()).await?;
 
-            // Fetch the mempool data for the block template:
-            // - if the mempool transactions change, we might return from long polling.
-            let mempool_txs = fetch_mempool_transactions(mempool).await?;
+                // Fetch the mempool data for the block template:
+                // - if the mempool transactions change, we might return from long polling.
+                //
+                // TODO:
+                // - add a `MempoolChange` type with an `async changed()` method.
+                // - if we are long polling, pause between state and mempool,
+                //   to allow transactions to re-verify (only works after PR #5841)
+                let mempool_txs = fetch_mempool_transactions(mempool.clone()).await?;
+
+                // - Long poll ID calculation
+                let server_long_poll_id = LongPollInput::new(
+                    chain_tip_and_local_time.tip_height,
+                    chain_tip_and_local_time.tip_hash,
+                    chain_tip_and_local_time.max_time,
```

Comment:
Do we want to use the max time from the client's long poll id?

Reply:
I think that would make the next poll ignore any changes in the block header due to a new `max_time` from the state?

The `max_time` can change if the local clock is close to or past the minimum difficulty time on testnet. (It doesn't matter on mainnet.)

Reply:
All the gory details are here:
https://github.com/ZcashFoundation/zebra/blob/4fb2417adc70f57f42e9df4c8467279058b743d7/zebra-state/src/service/read/difficulty.rs#L263-L276